pyomo/core/base/param.py

```diff
@@ -218,6 +223,9 @@ def __init__(self, *args, **kwd):
         self._mutable       = kwd.pop('mutable', Param.DefaultMutable )
         self._default_val   = kwd.pop('default', _NotValid )
         self._dense_initialize = kwd.pop('initialize_as_dense', False)
+        self._units         = kwd.pop('units', None)                                                                           
+        if self._units is not None:                                                                                            
+            self._mutable = True
```

Comment:
Should this generate a warning?  That is, should users be told that if they are using units, they *must* use mutable Params?

Reply:
Not sure. Is the main concern that we would be making everything immutable without them knowing?
Note: We are going to have to do some performance testing with this as well to see if the model build times change significantly.

Reply:
Right.  Mutable Params are measurably slower than immutable: they have to be constructed, and thy occupy a position in the expression tree (i.e., we cannot simplify them out).  I wonder if it is "evil" that we silently switch a user's Param if they happen to provide units.  A warning would force them to make it explicit, and could help educate them as to why "I added units to my model and suddenly everything is slower!"

I think what I was thinking was something along the lines of:
```python
if self._units is not None and not self._mutable:
    self._mutable = True
    logger.warning("Converting Param %s from immutable to mutable: Params with units must be mutable." % (self.name,))
```